### PR TITLE
piksi_system_daemon: Add async-child module.

### PR DESCRIPTION
**Pull request by gsmcmullin**
This adds a new module to piksi_system_daemon for management of asynchronous child processes with piped output.

A function `async_spawn` is provided to spawn such a process, and it takes callback function pointers for notification of the child's standard output and exit status.  The output is assumed to be text, and is split on newlines.  Both callback functions will be called from the zmq loop provided.  If needed in future it would be simple to add pipes for the child's stdin and stderr.

Internally, a pipe is used to syncronise the SIGCHLD handler where the child is reaped with the zmq loop from which the exit notification callback must be called.  I couldn't find a simpler safe way to do this.

This new mechanism is used for the SBP command implementation, instead of `popen`/`pclose` that were being used before.

This addresses issue https://github.com/swift-nav/firmware_team_planning/issues/248, and with these changes I have been able to successfully update the firmware, tested 10 times with a Linux host, and once with Windows.  I've also tested that some deliberate failures do fail correctly (empty image file, and corrupt image file).

/cc @jacobmcnamee @denniszollo

### CODE REVIEW

File: package/piksi_system_daemon/src/async-child.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        break;
+      }
+    }
+  }
+  free(ctx);
+
+  return 0;
+}
+
+int async_spawn(zloop_t *loop, char **argv,
```
Comment (gsmcmullin) [inline package/piksi_system_daemon/src/async-child.c]: Is this `argv` desirable, or is it better to take a single string command line and run with `sh -c`?

File: package/piksi_system_daemon/src/main.c [hunk 2]
```diff
@@ -361,15 +341,11 @@ static void sbp_command(u16 sender_id, u8 len, u8 msg_[], void* context)
...
-  ctx->pipe = popen(msg->command, "r");
   ctx->sequence = msg->sequence;
   ctx->pubsub_ctx = pubsub_ctx;
-  ctx->pollitem.fd = fileno(ctx->pipe);
-  ctx->pollitem.events = ZMQ_POLLIN;
-  int arg = O_NONBLOCK;
-  fcntl(fileno(ctx->pipe), F_SETFL, &arg);
-  zloop_poller(sbp_zmq_pubsub_zloop_get(ctx->pubsub_ctx),
-               &ctx->pollitem, command_output, ctx);
+  char *argv[] = {"upgrade_tool", "--debug", "upgrade.image_set.bin", NULL};
```
Comment (gsmcmullin) [inline package/piksi_system_daemon/src/main.c]: I guess we should use the actual string that come in the SBP message.

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-182-g8146309
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-182-g8146309/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-182-g8146309/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-182-g8146309

**Comment by lbara** [comment]
For the understand I have so far of the project, I am good checking this in! I'm also testing it now

**Comment by denniszollo** [comment]
Can we put this against release branch as well please to support customers potentially?

File: package/piksi_system_daemon/src/Makefile [hunk 3]
```diff
@@ -1,5 +1,6 @@
 TARGET=piksi_system_daemon
 SOURCES= \
+	async-child.c \
```
Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/Makefile]: hyphens now, huh?

**Comment by lbara** [comment]
There are some loose ends here. I agree with @denniszollo that the python file should be renamed from bootload_v3.py to something like firmware_upgrade.py

I also agree with @denniszollo that the output can be cleared a bit.
"INFO ok" is not very informative, nor "erasing mtdx" is very informative for the user.

I'll try to collect all the requirements, but for now this is great, since it is working and I'd really like to see this PR merged.

**Comment by lbara** [comment]
For the record, it worked great on OSX on the first try with bootload_v3 and took approximately 22 minutes.

File: package/piksi_system_daemon/src/async-child.c [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  async_children = ctx;
+}
+
+void async_child_waitpid_handler(pid_t pid, int status)
+{
+  /* Called by SIGCHLD handler with status from waitpid.
+   * We save the status here, and will deal with it from the zmq loop by
+   * writing to the wakeup pipe.
+   */
+  for (struct async_child_ctx *c = async_children; c; c = c->next) {
```
Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/async-child.c]: Since this function is executed from the `SIGCHLD` handler, you should mask `SIGCHLD` while modifying the `async_chld_ctx` list from other parts of the code. E.g. https://github.com/swift-nav/piksi_buildroot/blob/master/package/piksi_system_daemon/src/ports.c#L151

File: package/piksi_system_daemon/src/async-child.c (hunk 1 above)
Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/async-child.c]: I don't see any reason to change it. I think it's fine.

File: package/piksi_system_daemon/src/async-child.c [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  if (pipe(pipefd)) {
+    perror("async_spawn pipe");
+    free(ctx);
+    return -1;
+  }
+
+  ctx->pid = fork();
+  if (ctx->pid == 0) {
+    /* Code executed by new child */
+    close(STDIN_FILENO);
```
Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/async-child.c]: you could also `close(pipefd[0])` here since it won't be used

File: package/piksi_system_daemon/src/async-child.c [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                void (*output_callback)(const char *buf, void *ctx),
+                void (*exit_callback)(int status, void *ctx),
+                void *external_context)
+{
+  if (!wakepipe[0]) {
+    /* Set up the wake up pipe if not done already. */
+    pipe(wakepipe);
+    wakeitem.fd = wakepipe[0];
+    wakeitem.events = ZMQ_POLLIN;
+    zloop_poller(loop, &wakeitem, async_cleanup_dead_child, NULL);
```
Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/async-child.c]: This assumes the same `zloop` is passed in each time.

File: package/piksi_system_daemon/src/Makefile (hunk 3 above)
Comment (gsmcmullin) [inline package/piksi_system_daemon/src/Makefile]: Is that bad?  I'm happy to change it if there are strong feelings.

Comment (jacobmcnamee) [inline package/piksi_system_daemon/src/Makefile]: It doesn't matter to me. I've just seen nothing but underscores for the longest time.

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-190-ga6065fa
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-190-ga6065fa/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-190-ga6065fa/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-190-ga6065fa

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-190-ga6065fa
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-190-ga6065fa/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-190-ga6065fa/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-190-ga6065fa
